### PR TITLE
.Net: Replacing `FunctionsView` class with List

### LINKED ISSUE #604: Potential concurrency bugs in FunctionsView
FunctionsView appears to be trying to be thread-safe, as its SemanticFunctions and NativeFunctions are both ConcurrentDictionary.  However, in that case there are a couple of concurrency bugs lurking.

First, while the dictionaries are ConcurrentDictionary, the values stored for each key are `List<T>`.  The code for AddFunction looks like this:
```C#
            if (!this.SemanticFunctions.ContainsKey(view.SkillName))
            {
                this.SemanticFunctions[view.SkillName] = new();
            }

            this.SemanticFunctions[view.SkillName].Add(view);
```
If multiple threads concurrently try to add a FunctionView for the same skill, they could end up both mutating a `List<T>` concurrently, which is not safe.

Second, with that same code, if multiple threads both try to add a FunctionView for the same skill, and if both see that the SemanticFunctions doesn't yet contain that skill name, they will race to create the list to store for that skill.  While the dictionary itself won't be corrupted from a data structure perspective, one thread might add its list and then proceed to add to the list, and the other thread might then (having already checked ContainsKey and found it to return false) overwrite the original list with a new one, there by losing the data the other thread wrote.  Such initialization should be done using either TryAdd or GetOrAdd so that the dictionary provides the synchronization to ensure nothing is overwritten.

These issues can easily be fixed by using GetOrAdd and locking on the `List<>`... except the collections are exposed publicly from the FunctionsView, which means anyone else accessing these would also need to employ the same synchronization scheme.  That suggests the API should be designed differently.

(And related to these being exposed publicly, the properties also have public setters... if someone were to set these concurrently with other operations being performed, that could lead to other failure modes, e.g. because lookups are being done first by calling ContainsKey and then separately by using the indexer, the collection could actually be changed to a completely different instance between those calls.)

cc: @dluc, @shawncal

### PR DESCRIPTION
**Pull request by shawncal**
### Motivation and Context
The `FunctionsView` class wraps two inner Dictionaries of `FunctionView` objects: Native and Semantic. Since we no longer differentiate between these two for any practical purpose, we could merge those into one list, and we no longer need a special class for that.

### Description

- Removing the different collections to store Semantic vs Native functions
- Removing `FunctionsView` collection
- From `SkillCollection.GetFunctionsView()`, returning an `IReadOnlyList<FunctionView>`
- Callers can then easily use LINQ to filter the resulting flat list, or group by skill name.
- Updating callers, tests

Resolves #604

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:
    - BREAKING CHANGE as it modifies the public API. Low impact.

### CODE REVIEW

**Review by lemillermicrosoft — APPROVED** [review]

File: dotnet/src/SemanticKernel.Abstractions/SkillDefinition/IReadOnlySkillCollection.cs [hunk 1]
```diff
@@ -48,8 +49,6 @@ public interface IReadOnlySkillCollection
     /// <summary>
     /// Get all registered functions details, minus the delegates
     /// </summary>
-    /// <param name="includeSemantic">Whether to include semantic functions in the list</param>
-    /// <param name="includeNative">Whether to include native functions in the list</param>
     /// <returns>An object containing all the functions details</returns>
-    FunctionsView GetFunctionsView(bool includeSemantic = true, bool includeNative = true);
+    IReadOnlyList<FunctionView> GetFunctionsView();
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/SkillDefinition/IReadOnlySkillCollection.cs:53]: This should be named `GetFunctionViews` as it returns a list of `FunctionView` instances. Current name reflects the old implementation.

File: dotnet/src/SemanticKernel.Abstractions/SkillDefinition/NullReadOnlySkillCollection.cs [hunk 2]
```diff
@@ -38,9 +39,9 @@ public bool TryGetFunction(string skillName, string functionName, [NotNullWhen(t
     }
 
     /// <inheritdoc/>
-    public FunctionsView GetFunctionsView(bool includeSemantic = true, bool includeNative = true)
+    public IReadOnlyList<FunctionView> GetFunctionsView()
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/SkillDefinition/NullReadOnlySkillCollection.cs:42]: As above this should be named `GetFunctionViews`

File: dotnet/src/SemanticKernel/SkillDefinition/SkillCollection.cs [hunk 3]
```diff
@@ -85,21 +85,15 @@ public bool TryGetFunction(string skillName, string functionName, [NotNullWhen(t
     }
 
     /// <inheritdoc/>
-    public FunctionsView GetFunctionsView(bool includeSemantic = true, bool includeNative = true)
+    public IReadOnlyList<FunctionView> GetFunctionsView()
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel/SkillDefinition/SkillCollection.cs:88]: Name change required here also

**Review by SergeyMenshykh — APPROVED** [review]

**Comment by RogerBarreto** [comment]
Resolves #604

File: dotnet/src/SemanticKernel.Abstractions/SkillDefinition/IReadOnlySkillCollection.cs (hunk 1 above)
Comment (shawncal) [reply dotnet/src/SemanticKernel.Abstractions/SkillDefinition/IReadOnlySkillCollection.cs:53]: Done

**Review by markwallace-microsoft — APPROVED** [review]

**Pull request merged**